src/software/ai/passing/receiver_position_generator.hpp

```diff
@@ -0,0 +1,344 @@
+#pragma once
+
+#include <iomanip>
+#include <random>
+
+#include "proto/message_translation/tbots_protobuf.h"
+#include "proto/parameters.pb.h"
+#include "software/ai/passing/cost_function.h"
+#include "software/ai/passing/field_pitch_division.h"
+#include "software/ai/passing/pass.h"
+#include "software/ai/passing/pass_with_rating.h"
+#include "software/logger/logger.h"
+#include "software/world/world.h"
+
+/**
+ * This class is responsible for generating the best positions for our pass
+ * receivers to go to
+ */
+template <class ZoneEnum>
+class ReceiverPositionGenerator
+{
+    static_assert(std::is_enum<ZoneEnum>::value,
+                  "PassGenerator: ZoneEnum must be a zone id enum");
+
+   public:
+    /**
+     * Creates a new ReceiverPositionGenerator
+     * @param pitch_division The pitch division to split the receivers into
+     * @param passing_config The passing configuration to use when looking for best
+     * receiving positions
+     */
+    explicit ReceiverPositionGenerator(
+        std::shared_ptr<const FieldPitchDivision<ZoneEnum>> pitch_division,
+        TbotsProto::PassingConfig passing_config);
+
+    /**
+     * Generates the best receiving positions for the friendly robots to go to
+     * @param world The world to generate the best receiving positions based on
+     * @param num_positions The number of receiving positions to generate
+     * @param existing_receiver_positions A set of existing receiver positions that will
+     * be avoided, if possible, when generating the new receiver positions.
+     * @param pass_origin_override An optional override for the position where the pass
+     * will be made from. If not provided, the ball position will be used. This could be
+     * helpful if you are trying to position the receivers based on the ball's future
+     * position.
+     * @return A vector of
+     *      min(num_positions, num_friendly_robots - existing_receiver_positions.size())
+     * positions that the receivers could use.
+     */
+    std::vector<Point> getBestReceivingPositions(
+        const World &world, unsigned int num_positions,
+        const std::vector<Point> &existing_receiver_positions = {},
+        const std::optional<Point> &pass_origin_override      = std::nullopt);
+
+
+   private:
+    /**
+     * Sample more points and update the best_receiving_positions map if better
+     * receiving positions in each zone were found
+     *
+     * @param best_receiving_positions The map of the best receiving positions for each
+     * zone found so far, and their ratings.
+     * @param world The world to sample receiving positions in
+     * @param zones_to_sample The subset of the zones to sample receiving positions in
```

Comment:
nit: param ordering doesn't match the function

Reply:
Updated